### DISCUSSION TITLE
How to implement websockets

### DISCUSSION THREAD

**Discussion by DenuxPlays** [actix/actix-web#3323] [Q&A] [answered]
Hey,

I wrote an API in actix and I also created and event system that we use for handling intern events. (for example if an entity was created etc.)
It is based on the Tokio sync channels.
I've wondered what is the best way to implement web sockets on top of it?
I saw a lot of implementations on top of actors but our event system is not actor based.
I also saw the actix-ws crate but it doesn't seem to be maintained anymore.

So is there any other technology to use (like socket.io) or basically how/what is the best/a clean way to implement web sockets?

Edit:
If you need more information the whole backend ist open source:
https://github.com/financrr/financrr-app

**Comment by robjtede** [comment, marked as answer]
The Actix team adopted `actix-ws` and it is is now back in development. I'd recommend using it, even in it's current state, but look out for updates soon.

**Reply by MeRahulAhire** [reply]
@robjtede Is actix_ws production ready and what's the current state of it? I'm also trying to understand actix_ws from last few days but because there's little to no examples in the docs I'm struggling to understand it unlike socket.io which is literally copy and paste in my humble opinion. Do you know any resource that would help me understand it like creating a global live connection and then in post routes or any other function we can emit the event continuously?

**Reply by robjtede** [reply]
The -actorless projects in the examples repo showcase the use of actix-ws: https://github.com/actix/examples/tree/master/websockets

The library is production quality, and has surpassed the old WS library in terms of features since the OP was written.

socket.io is an opinionated layer onto of websocket libraries. actix-ws itself does not try to implement concepts such as rooms.
